New Serverless Pattern - Add Lambda Durable Functions - Human Approval Pattern (#2874)
aws-samples/serverless-patterns: lambda-durable-human-approval-sam/README.md

# Human-in-the-Loop Approval Workflow with AWS Lambda durable functions

This pattern demonstrates a human-in-the-loop approval workflow using AWS Lambda durable functions. The workflow pauses execution for up to 24 hours while waiting for human approval via email, automatically handling timeouts and resuming when decisions are received.

> **Important:** Lambda durable functions have limited regional availability. Please check the [AWS documentation](https://docs.aws.amazon.com/lambda/latest/dg/durable-functions.html) for current regional support.

Reviewer: ```suggestion
**Important:** Please check the [AWS documentation](https://docs.aws.amazon.com/lambda/latest/dg/durable-functions.html) for regions currently supported by AWS Lambda durable functions.
```

> # Human-in-the-Loop Approval Workflow with Lambda Durable Functions

Author: corrected

> This pattern demonstrates a human-in-the-loop approval workflow using AWS Lambda Durable Functions. The workflow pauses execution for up to 24 hours while waiting for human approval via email, automatically handling timeouts and resuming when decisions are received.

Author: corrected

    approve_url = f"{api_base_url}/approve/{request_uuid}"
    
    # Wait for callback result
    approval = callback.result()
```

### Callback Handler

When user clicks approve/reject:

```python
def lambda_handler(event, context):
    # Get UUID from URL path
    request_uuid = event['pathParameters']['uuid']
    
    # Fetch callback ID from DynamoDB
    response = table.get_item(Key={'uuid': request_uuid})
    callback_id = response['Item']['callbackId']
    
    # Send callback to durable function
    lambda_client.send_durable_execution_callback_success(
        CallbackId=callback_id,
        Result=json.dumps({'decision': 'approved'})
    )
```

### Cost Efficiency

**Traditional Lambda approach:**
- Cannot wait more than 15 minutes (Lambda max timeout)
- Would need Step Functions or polling mechanism
- Complex state management

**Durable Functions approach:**
- Wait up to 24 hours
- No compute charges during wait
- Simple, clean code
- Automatic state management

## Configuration

### Adjust Timeout Duration

Modify in `template.yaml`:

```yaml
DurableConfig:
  ExecutionTimeout: 86400  # 24 hours in seconds
```

And in `src/lambda_function.py`:

```python

> config=CallbackConfig(timeout=Duration.from_hours(24))

Author: done

```

### Change Email Address

Update during deployment or redeploy with new email:

```bash
sam deploy --parameter-overrides ApproverEmail=new-email@example.com
```

## Use Cases

- **Expense Approvals** - Wait for manager approval on spending requests
- **Document Reviews** - Pause workflow while documents are reviewed
- **Deployment Approvals** - Require human approval before production deployments
- **Access Requests** - Pause while security team reviews access requests
- **Contract Approvals** - Wait for legal team approval on contracts

## Monitoring

### CloudWatch Logs

Monitor the durable function:
```bash
aws logs tail /aws/lambda/lambda-durable-hitl-approval-ApprovalFunction-XXXXX \
  --region us-east-2 \
  --follow
```

Monitor the callback handler:
```bash
aws logs tail /aws/lambda/lambda-durable-hitl-approv-CallbackHandlerFunction-XXXXX \
  --region us-east-2 \
  --follow
```

> ### DynamoDB Table

Author: removed
